mesonbuild/backend/ninjabackend.py

```diff
@@ -1926,6 +1926,33 @@ def generate_rust_target(self, target: build.BuildTarget) -> None:
                 # Rust uses -l for non rust dependencies, but we still need to
                 # add dylib=foo
                 args += ['-l', f'dylib={d.name}']
+
+        # Since 1.61.0 Rust has a special modifier for whole-archive linking,
+        # before that it would treat linking two static libraries as
+        # whole-archive linking.
+        if mesonlib.version_compare(rustc.version, '>= 1.61.0'):
+            whole_archive = ':+whole-archive'
+        else:
+            whole_archive = ''
+
+        for d in target.link_whole_targets:
+            linkdirs.add(d.subdir)
+            if d.uses_rust():
+                # specify `extern CRATE_NAME=OUTPUT_FILE` for each Rust
+                # dependency, so that collisions with libraries in rustc's
+                # sysroot don't cause ambiguity
+                args += ['--extern', '{}={}'.format(d.name, os.path.join(d.subdir, d.filename))]
+                project_deps.append(RustDep(d.name, self.rust_crates[d.name].order))
+            else:
```

Comment:
This else conditional was not being exercised in the tests, so I pushed a commit that exercises it, and it fails:

```
[1/5] ccache cc -Ilibvalue.a.p -I. -I.. -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -fPIC -MD -MQ libvalue.a.p/value.c.o -MF libvalue.a.p/value.c.o.d -o libvalue.a.p/value.c.o -c ../value.c
[2/5] rm -f libvalue.a && gcc-ar csrDT libvalue.a libvalue.a.p/value.c.o
[3/5] rustc -C linker=cc --color=always --crate-type rlib -g -C relocation-model=pic --crate-name other --emit dep-info=other.d --emit link -o libother.rlib ../other.rs
[4/5] rustc -C linker=cc --color=always --crate-type rlib -g -C relocation-model=pic --crate-name stuff --emit dep-info=stuff.d --emit link -o libstuff.rlib --extern other=libother.rlib -l static:+whole-archive=value -L . ../stuff.rs
FAILED: libstuff.rlib 
rustc -C linker=cc --color=always --crate-type rlib -g -C relocation-model=pic --crate-name stuff --emit dep-info=stuff.d --emit link -o libstuff.rlib --extern other=libother.rlib -l static:+whole-archive=value -L . ../stuff.rs
error: the linking modifiers `+bundle` and `+whole-archive` are not compatible with each other when generating rlibs

error: aborting due to previous error
```